streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -368,6 +370,7 @@ public void setUncaughtExceptionHandler(final Thread.UncaughtExceptionHandler eh
      * @param queryableStoreType    accept only stores that are accepted by {@link QueryableStoreType#accepts(StateStore)}
      * @param <T>                   return type
      * @return  A facade wrapping the {@link org.apache.kafka.streams.processor.StateStore} instances
+     * @throws org.apache.kafka.streams.errors.InvalidStateStoreException if the streams are (re-)initializing
```

Comment:
Actually, thinking about it a bit more i think that for point 2 we do actually want to do the empty check and throw `InvalidStateStoreException`. Why? if there was a rebalance and the store migrated elsewhere, but this call happens after the migration and `allStores.isEmpty()` then this indicates that the store is no longer on this instance, so we should throw.
i.e,:
```
store = streams.store(...); // store exists at this point and not rebalancing
// some time passes and a rebalance happen, store gets migrated elsewhere
store.get("blah");  // should throw InvalidStateStoreException as the store no longer exists in this instance